File: kotlin-node/karakum/src/jsMain/resources/annotations/annotateConflictingEntityNames.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

const conflictingEntities = {
    "inspector": [
        "Session",
    ],
    "readline": [
        "Interface",
    ],
}

function checkNameByNamespace(node, context) {
    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
    const namespace = typeScriptService?.findClosest(node, ts.isModuleDeclaration)
    if (!namespace) return false

    const namespaceConflictingEntities = conflictingEntities[namespace.name.text]
    if (!namespaceConflictingEntities) return false

    return namespaceConflictingEntities.includes(node.name?.text);
}

function checkNameByFileName(node, context) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

    const [, fileNameConflictingEntities] = Object.entries(conflictingEntities)
        .find(([fileName]) => sourceFileName.endsWith(fileName)) ?? []

    if (!fileNameConflictingEntities) return false

    return fileNameConflictingEntities.includes(node.name?.text);
}

export default (node, context) => {
    if (
        ts.isClassDeclaration(node)
        && node.name
        && (
            checkNameByNamespace(node, context)
            || checkNameByFileName(node, context)
        )
    ) {
        return `@JsName("${node.name.text}")`
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/annotations/annotateConflictingEntityNames.kt
package node.karakum.annotations

import io.github.sgrishchenko.karakum.extension.AnnotationContext
import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import arrow.core.raise.nullable
import typescript.Node
import typescript.isClassDeclaration
import typescript.isStringLiteral

private val conflictingEntities = mapOf(
    "inspector" to setOf(
        "Session",
    ),
    "readline" to setOf(
        "Interface",
    ),
)

private fun checkNameByNamespace(node: Node, context: Context) = nullable {
    ensure(isClassDeclaration(node))

    val typeScriptService = ensureNotNull(context.lookupService(typeScriptServiceKey))
    val namespace = ensureNotNull(typeScriptService.findClosestNamespace(node))

    val namespaceName = namespace.name
    ensure(isStringLiteral(namespaceName))

    val namespaceConflictingEntities = ensureNotNull(conflictingEntities[namespaceName.text])
    ensure(node.name?.text in namespaceConflictingEntities)
} != null

private fun checkNameByFileName(node: Node) = nullable {
    ensure(isClassDeclaration(node))

    val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
    val (_, fileNameConflictingEntities) = conflictingEntities.entries
        .find { (fileName) -> sourceFileName.endsWith(fileName) }
        .let { ensureNotNull(it) }
    ensure(node.name?.text in fileNameConflictingEntities)
} != null

fun annotateConflictingEntityNames(node: Node, context: AnnotationContext) = nullable {
    ensure(isClassDeclaration(node))
    ensure(
        checkNameByNamespace(node, context)
                || checkNameByFileName(node)
    )

    val name = ensureNotNull(node.name)

    "@JsName(\"${name.text}\")"
}
